[LENS-414] Confirm modal style improvements

dbchristopher opened the pull request:
### :sparkles: Changes

- Update base Confirm text styling to match [figma](https://www.figma.com/file/sDUnKT8DfIroF2yNbXfrc2/Dialog?node-id=168%3A1667), including a more reasonable in and max width
- Add support for using a React component as the modal message, including documentation for rich content example.
-Update base theme to render more vibrant button colors

### :white_check_mark: Requirements

- [x] Includes test coverage for all changes
- [x] Build and tests are passing
- [x] PR is ideally < 400LOC
- [x] Link(s) to related Github issues

### :camera: Screenshots

<img width="823" alt="Screen Shot 2019-11-20 at 11 54 00 AM" src="https://user-images.githubusercontent.com/238293/69273027-8309b580-0b8c-11ea-8f8c-ac0540c310c8.png">
<img width="815" alt="Screen Shot 2019-11-20 at 11 53 53 AM" src="https://user-images.githubusercontent.com/238293/69273028-8309b580-0b8c-11ea-9526-4b3cd58b6f4e.png">

dbchristopher commented on packages/components/src/Modal/Dialog/DialogSurface.tsx:
```diff
@@ -40,7 +40,4 @@ export const DialogSurface = styled(ModalSurface)<ModalSurfaceProps>`
-  maxHeight: '90vh',
-  maxWidth: '100%',
-}
+DialogSurface.defaultProps = {}
```
Moved these default props to ModalSurface

lukelooker replied:
Since it's empty probably best to just go ahead and remove this altogether :)

lukelooker commented on packages/components/src/Modal/Layout/ModalFooter.tsx:
```diff
@@ -53,13 +53,7 @@ const Layout = styled.footer`
-    * so that the content is properly aligned.
-    * padding is: vertical: large, horizontal: calc(xlarge - xsmall)
-    **/
+  padding: ${({ theme }) => `${theme.space.large} ${theme.space.xlarge}`};
```
This change causes a couple of issues:

1. There's margin on the left hand side of the secondary button(s)
2. The gap between the buttons is narrower (previously it was `xsmall + xsmall` whereas now it's just `xsmall`)

![image](https://user-images.githubusercontent.com/34253496/69290910-481c7780-0bb6-11ea-89a9-ea39674a75a6.png)

dbchristopher replied:
Ok I updated the footer layout to use `grid-gap` to handle that item spacing, rather than margins. That way we don't have to deal with any finicky `calc` subtraction to get footer items to line up with content

lukelooker replied:
Awesome

lukelooker commented on packages/components/src/Modal/Layout/ModalHeader.tsx:
```diff
@@ -58,31 +59,27 @@ export const ModalHeader: FC<ModalHeaderProps> = ({
-  align-items: center;
-  flex-direction: row;
-
-  ${Heading} {
```
This needs to be restored (for now at least).

It reconciles a bug where we previously mis-calculated Heading if no `as` or `fontSize` was explicitly specified. Changing this is definitely something we should probably explore but would need an accomodating fix within Helltool to prevent changes to presentation and snapshots.

dbchristopher replied:
So assuming I can update helltool to strip out now-redundant inline `<Heading>` components, no change is needed here right? As we'd now be providing the proper formatting out of the box I don't think we'd have a case where a heading was used without `as` or `fontSize`

lukelooker replied:
Yep - that'd be great. Let's make sure to note that in the CHANGELOG since it's a change in behavior (albeit essentially a bugfix)

lukelooker replied:
That works for me!

lukelooker commented on packages/components/src/Modal/Layout/ModalHeader.tsx:
```diff
@@ -58,31 +59,27 @@ export const ModalHeader: FC<ModalHeaderProps> = ({
+          icon={closeIcon}
+          mr="-10px"
+        />
+      </Flex>
```
This changes makes me very happy but there's a couple of minor nits.

A lot of existing implementations already have a `<Heading>` component within them so we'll generate a bunch of errors where React is producing `Heading > Heading`

IconButton - the size `medium` is a bit larger than the old version. `small` would produce a `20x20` SVG which would be inline with the previous version.

dbchristopher replied:
Looks like there are only 6 instances of `<ModalHeader>` in Helltool at the moment -- I'm happy to take point on migrating those! I do feel like only accepting `string` for that specific value is the right move either way

Relatedly, there are exactly zero uses of `<Confirm>` in helltool at the moment 😖 Hopefully the improved style and more usable maxwidth will encourage adoption there

dbchristopher replied:
> IconButton - the size medium is a bit larger than the old version. small would produce a 20x20 SVG which would be inline with the previous version.

Yeah it is bigger than the old one for sure. But it also seems closer to the figma reference to my eyes. Although the figma layout certainly lands somewhere in the middle between those two sizes 😖 What do you think?

<img width="811" alt="Screen Shot 2019-11-21 at 3 31 11 PM" src="https://user-images.githubusercontent.com/238293/69385450-3ac5c280-0c74-11ea-852a-afdaf567fe9a.png">

dbchristopher replied:
Also you'll notice I did allow the max-width to go a little wider than what's in figma, as I don't think the modal-footer example with four actions would fit in that narrower layout:

<img width="734" alt="Screen Shot 2019-11-21 at 3 35 09 PM" src="https://user-images.githubusercontent.com/238293/69385535-91330100-0c74-11ea-9ba0-f3a5583d1172.png">

Do you think the 700px max width is the right choice? Or should I also reduce that to conform closer to what's in figma?

jhardy replied:
I am working on having a better spec in figma for some of these items. I would say for now we can go with what looks good as the current spec is a bit outdated and need some refinement based on some newer patterns we have in the codebase

lukelooker replied:
Works for me. 

Just to clarify the example is slightly contrived since I've never seen a real-world implementation (thus-far) that has more than 1 "secondary" button :)

lukelooker commented on packages/www/src/documentation/components/modals/confirm.mdx:
````diff
@@ -60,3 +60,78 @@ A render-props version, for use in class components.
+    </Confirm>
+  )
+}
+```
````
I'd propose a somewhat simpler example (I'd prefer not so show `styled` in examples except where we're explicitly try to teach folks about that functionality):

```

```jsx
;() => {
  /*
   * import { Confirm, Flex, FlexItem, Icon, Paragraph, Link } from '@looker/components'
   * import styled from 'styled-components'
   */

  const RichMessage = () => (
    <Flex>
      <FlexItem>
        <Icon
          mr="large"
          name="CircleInfoOutline"
          size="80px"
          color="palette.purple400"
        />
      </FlexItem>
      <FlexItem>
        <Paragraph>
          Canadians say "sorry" so often that{' '}
          <strong>in 2009 a law was passed</strong> declaring that an apology
          cannot be used as an admission of guilt.
        </Paragraph>
        <Cite>
          SOURCE:{' '}
          <Link
            href="https://www.theloop.ca/canadians-love-to-say-sorry-so-much-we-had-to-make-this-law/"
            target="_blank"
          >
            the loop
          </Link>
        </Cite>
      </FlexItem>
    </Flex>
  )

  const Cite = styled.cite`
    font-size: ${({ theme }) => theme.fontSizes.small};
    color: ${({ theme }) => theme.colors.palette.charcoal300};
    margin-top: ${({ theme }) => theme.space.xsmall};
    display: block;
  `

  return (
    <Confirm
      title="Did you know?"
      message={<RichMessage />}
      onConfirm={close => {
        alert('Now you know.')
        close()
      }}
    >
      {open => <Button onClick={open}>Do something fancy</Button>}
    </Confirm>
  )
}
```

dbchristopher replied:
Great simplification 👍

dbchristopher commented:
Ok @lukelooker I addressed all the feedback and this should be ready for another look! It still includes the theme updates, so once #113 is merged in the line diffs should be greatly reduced

dbchristopher commented on packages/components/src/Icon/Icon.tsx:
```diff
@@ -70,6 +70,10 @@ const Styled = styled.div<Omit<IconProps, 'name'>>`
+
+  svg {
+    width: inherit;
+  }
```
Fixes an issue where the icon gets rendered super tiny for some reason when used in www
<img width="466" alt="Screen Shot 2019-11-25 at 11 45 48 AM" src="https://user-images.githubusercontent.com/238293/69574519-399adb00-0f7d-11ea-940b-67d58701dbd6.png">

dbchristopher commented:
ok @lukelooker I think I addressed all the feedback, and fixed the bug in the Drawer layout. This should be ready for another look whenever you're around! No rush 😄

dbchristopher commented:
Awesome, thanks @lukelooker!
